File: docs/examples/grpc_client.py
"""Example gRPC client for Flask backend services.

This example demonstrates how to communicate with gRPC services
exposed by the Flask backend on port 50051.
"""

import grpc
import os
import sys
from datetime import datetime

# Add proto package to path
sys.path.insert(
    0, os.path.join(os.path.dirname(__file__), "../../services/flask-backend/app/grpc")
)

try:
    from protos import template_pb2, template_pb2_grpc
except ImportError:
    print("Error: Proto files not compiled")
    print("Run: python -m grpc_tools.protoc -I./services/flask-backend/app/grpc/protos")
    print("       --python_out=./services/flask-backend/app/grpc")
    print("       --grpc_python_out=./services/flask-backend/app/grpc")
    print("       ./services/flask-backend/app/grpc/protos/template.proto")
    sys.exit(1)


class GRPCClient:
    """Client for gRPC services."""

    def __init__(self, host="localhost", port=50051):
        self.channel = grpc.insecure_channel(f"{host}:{port}")
        self.auth_stub = template_pb2_grpc.AuthServiceStub(self.channel)
        self.team_stub = template_pb2_grpc.TeamServiceStub(self.channel)
        self.health_stub = template_pb2_grpc.HealthStub(self.channel)

    def validate_token(self, token):
        """Validate JWT token."""
        print(f"Validating token: {token[:20]}...")
        request = template_pb2.TokenRequest(token=token)
        response = self.auth_stub.ValidateToken(request)

        if response.valid:
            print("✓ Token valid")
            print(f"  User: {response.user_id}")
            print(f"  Role: {response.role}")
            print(f"  Teams: {', '.join(response.team_ids) or 'none'}")
        else:
            print(f"✗ Token invalid: {response.error}")

        return response

    def get_user(self, user_id):
        """Get user by ID."""
        print(f"Getting user: {user_id}")
        request = template_pb2.UserRequest(user_id=user_id)
        response = self.auth_stub.GetUser(request)

        if response.error:
            print(f"✗ Error: {response.error}")
        else:
            print("✓ User found")
            print(f"  ID: {response.id}")
            print(f"  Email: {response.email}")
            print(f"  Name: {response.full_name}")
            print(f"  Role: {response.role}")

        return response

    def get_team(self, team_id):
        """Get team by ID."""
        print(f"Getting team: {team_id}")
        request = template_pb2.TeamRequest(team_id=team_id)
        response = self.team_stub.GetTeam(request)

        if response.error:
            print(f"✗ Error: {response.error}")
        else:
            print("✓ Team found")
            print(f"  ID: {response.id}")
            print(f"  Name: {response.name}")
            print(f"  Slug: {response.slug}")
            print(f"  Owner: {response.owner_id}")

        return response

    def list_user_teams(self, user_id):
        """List teams for user."""
        print(f"Listing teams for user: {user_id}")
        request = template_pb2.UserRequest(user_id=user_id)
        response = self.team_stub.ListUserTeams(request)

        if response.error:
            print(f"✗ Error: {response.error}")
        else:
            print(f"✓ Found {response.total} teams")
            for team in response.teams:
                print(f"  - {team.name} ({team.slug})")

        return response

    def health_check(self):
        """Check server health."""
        print("Checking health...")
        from google.protobuf.empty_pb2 import Empty

        request = Empty()
        response = self.health_stub.Check(request)

        if response.status == template_pb2.HealthCheckResponse.SERVING:
            print("✓ Server is serving")
        else:
            print("✗ Server is not serving")

        return response

    def close(self):
        """Close connection."""
        self.channel.close()


def main():
    """Example usage."""
    print("=== gRPC Client Example ===\n")

    # Create client
    client = GRPCClient(host="localhost", port=50051)

    try:
        # Health check
        client.health_check()
        print()

        # Validate token (replace with actual token)
        client.validate_token("example_jwt_token_here")
        print()

        # Get user
        client.get_user("user_123")
        print()

        # Get team
        client.get_team("team_456")
        print()

        # List user teams
        client.list_user_teams("user_123")
        print()

    except grpc.RpcError as e:
        print(f"gRPC error: {e.code()} - {e.details()}")
    except Exception as e:
        print(f"Error: {e}")
    finally:
        client.close()


if __name__ == "__main__":
    main()

File: docs/examples/grpc_client.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/protobuf/types/known/emptypb"

	pb "github.com/penguintechinc/project-template/services/go-backend/internal/grpc/protos"
)

// GRPCClient wraps gRPC service clients
type GRPCClient struct {
	conn      *grpc.ClientConn
	authCli   pb.AuthServiceClient
	teamCli   pb.TeamServiceClient
	healthCli pb.HealthClient
}

// NewGRPCClient creates a new gRPC client
func NewGRPCClient(host string, port int) (*GRPCClient, error) {
	addr := fmt.Sprintf("%s:%d", host, port)
	conn, err := grpc.NewClient(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, fmt.Errorf("failed to connect to %s: %w", addr, err)
	}

	return &GRPCClient{
		conn:      conn,
		authCli:   pb.NewAuthServiceClient(conn),
		teamCli:   pb.NewTeamServiceClient(conn),
		healthCli: pb.NewHealthClient(conn),
	}, nil
}

// ValidateToken validates JWT token
func (c *GRPCClient) ValidateToken(token string) (*pb.TokenResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	fmt.Printf("Validating token: %s...\n", token[:20])
	resp, err := c.authCli.ValidateToken(ctx, &pb.TokenRequest{
		Token: token,
	})
	if err != nil {
		return nil, err
	}

	if resp.Valid {
		fmt.Println("✓ Token valid")
		fmt.Printf("  User: %s\n", resp.UserId)
		fmt.Printf("  Role: %s\n", resp.Role)
		fmt.Printf("  Teams: %v\n", resp.TeamIds)
	} else {
		fmt.Printf("✗ Token invalid: %s\n", resp.Error)
	}

	return resp, nil
}

// GetUser gets user by ID
func (c *GRPCClient) GetUser(userID string) (*pb.UserResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	fmt.Printf("Getting user: %s\n", userID)
	resp, err := c.authCli.GetUser(ctx, &pb.UserRequest{
		UserId: userID,
	})
	if err != nil {
		return nil, err
	}

	if resp.Error == "" {
		fmt.Println("✓ User found")
		fmt.Printf("  ID: %s\n", resp.Id)
		fmt.Printf("  Email: %s\n", resp.Email)
		fmt.Printf("  Name: %s\n", resp.FullName)
		fmt.Printf("  Role: %s\n", resp.Role)
	} else {
		fmt.Printf("✗ Error: %s\n", resp.Error)
	}

	return resp, nil
}

// GetTeam gets team by ID
func (c *GRPCClient) GetTeam(teamID string) (*pb.TeamResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	fmt.Printf("Getting team: %s\n", teamID)
	resp, err := c.teamCli.GetTeam(ctx, &pb.TeamRequest{
		TeamId: teamID,
	})
	if err != nil {
		return nil, err
	}

	if resp.Error == "" {
		fmt.Println("✓ Team found")
		fmt.Printf("  ID: %s\n", resp.Id)
		fmt.Printf("  Name: %s\n", resp.Name)
		fmt.Printf("  Slug: %s\n", resp.Slug)
		fmt.Printf("  Owner: %s\n", resp.Owner)
	} else {
		fmt.Printf("✗ Error: %s\n", resp.Error)
	}

	return resp, nil
}

// ListUserTeams lists teams for user
func (c *GRPCClient) ListUserTeams(userID string) (*pb.TeamListResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	fmt.Printf("Listing teams for user: %s\n", userID)
	resp, err := c.teamCli.ListUserTeams(ctx, &pb.UserRequest{
		UserId: userID,
	})
	if err != nil {
		return nil, err
	}

	if resp.Error == "" {
		fmt.Printf("✓ Found %d teams\n", resp.Total)
		for _, team := range resp.Teams {
			fmt.Printf("  - %s (%s)\n", team.Name, team.Slug)
		}
	} else {
		fmt.Printf("✗ Error: %s\n", resp.Error)
	}

	return resp, nil
}

// HealthCheck checks server health
func (c *GRPCClient) HealthCheck() (*pb.HealthCheckResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	fmt.Println("Checking health...")
	resp, err := c.healthCli.Check(ctx, &emptypb.Empty{})
	if err != nil {
		return nil, err
	}

	if resp.Status == pb.HealthCheckResponse_SERVING {
		fmt.Println("✓ Server is serving")
	} else {
		fmt.Println("✗ Server is not serving")
	}

	return resp, nil
}

// Close closes the connection
func (c *GRPCClient) Close() error {
	return c.conn.Close()
}

func main() {
	fmt.Println("=== gRPC Client Example ===\n")

	client, err := NewGRPCClient("localhost", 50052)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	defer client.Close()

	// Health check
	_, err = client.HealthCheck()
	if err != nil {
		log.Printf("Health check error: %v", err)
	}
	fmt.Println()

	// Validate token
	_, err = client.ValidateToken("example_jwt_token_here")
	if err != nil {
		log.Printf("Validate token error: %v", err)
	}
	fmt.Println()

	// Get user
	_, err = client.GetUser("user_123")
	if err != nil {
		log.Printf("Get user error: %v", err)
	}
	fmt.Println()

	// Get team
	_, err = client.GetTeam("team_456")
	if err != nil {
		log.Printf("Get team error: %v", err)
	}
	fmt.Println()

	// List user teams
	_, err = client.ListUserTeams("user_123")
	if err != nil {
		log.Printf("List teams error: %v", err)
	}
}
